### PR TITLE
Refactor `retro()`, clean up iterators

### PR DESCRIPTION
**Pull request by afshin**
Follow on to https://github.com/jupyterlab/lumino/pull/346

### CODE REVIEW

**Comment by afshin** [comment]
The [RTD failure](https://readthedocs.org/projects/lumino/builds/17848344/) seems to be unrelated:
> ```
> UnavailableInvalidChannel: The channel is not accessible or is invalid.
>   channel name: pkgs/main
>   channel url: https://repo.anaconda.com/pkgs/main
>   error code: 404```

**Review by blink1073 — APPROVED** [review]
LGTM!  No idea what is going on with RTD.

**Comment by blink1073** [comment]
Same thing is happening in `jupyterlab_server`: https://readthedocs.org/projects/jupyterlab-server/builds/17855741/

**Comment by blink1073** [comment]
Sigh.  https://github.com/readthedocs/readthedocs.org/issues/9527

**Review by fcollonval — APPROVED** [review]
LGTM

File: packages/algorithm/src/retro.ts [hunk 1]
```diff
@@ -38,15 +38,18 @@ export interface IRetroable<T> {
...
   if (typeof (object as any).retro === 'function') {
-    return (object as IRetroable<T>).retro();
-  }
-  return (function* (object) {
-    for (let index = object.length - 1; index > -1; index--) {
-      yield object[index];
+    const it = (object as IRetroable<T>).retro();
+    let item: IteratorResult<T>;
+    while (!(item = it.next()).done) {
+      yield item.value;
```
Comment (vidartf) [inline packages/algorithm/src/retro.ts:48]: FYI, you could have also used `yield*` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield*

**Pull request merged**